src/parallel/parallel.cpp

```diff
@@ -123,32 +123,192 @@ ThreadManager::get_segment_index(size_t index)
+////////////////////////////////////////////////////////////////////////////////
+bool
+ParallelBucket::is_empty() 
+{
+    if (this->offset == SIZE_MAX) {
+        return true;
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// STATIC HELPER FUNCTIONS
+////////////////////////////////////////////////////////////////////////////////
+
+HashCodeType
+hash(const KeyType key) {
+  LOG_TRACE("Enter");
+  size_t k = static_cast<size_t>(key);
+  // I use the suffix '*ULL' to denote that the literal is at least an int64.
+  // I've had weird bugs in the past to do with literal conversion. I'm not sure
+  // the details. I only remember it was a huge pain.
+  k = ((k >> 30) ^ k) * 0xbf58476d1ce4e5b9ULL;
+  k = ((k >> 27) ^ k) * 0x94d049bb133111ebULL;
+  k =  (k >> 31) ^ k;
+  // We could theoretically use `reinterpret_cast` to ensure there is no
+  // overhead in this cast because HashCodeType is typedef'ed to size_t.
+  return static_cast<HashCodeType>(k);
+}
+
+size_t
+get_home(const HashCodeType hashcode, const size_t capacity) {
+  LOG_TRACE("Enter");
+  size_t h = static_cast<size_t>(hashcode);
+  return h % capacity;
+}
+
 ////////////////////////////////////////////////////////////////////////////////
 /// HASH TABLE
 ////////////////////////////////////////////////////////////////////////////////
 
 bool
 ParallelRobinHoodHashTable::insert(KeyType key, ValueType value)
 {
-    // TODO
+    HashCodeType hashcode = hash(key);
+    size_t home = get_home(hashcode, this->capacity_);
+    auto [is_inserted, key_exists] = distance_zero_insert(key, value, home);
+    if(is_inserted) {
+        return true;
+    }
+    if(key_exists) {
+        return false;
+    }
+    ThreadManager manager = this->get_thread_lock_manager();
+
+    size_t offset = 0;
+    ParallelBucket entry_to_swap = {.key = key, .value = value, .hashcode = hashcode, .offset = offset};
+
+    auto [next_index, found] = this->find_next_index_lock(manager, home, key, offset);
+
+    if (found) {
+        manager.release_all_locks();
+        return false;
+    }
+
+    locked_insert(entry_to_swap, next_index);
+    manager.release_all_locks();
+    return true;
 }
 
 void
 ParallelRobinHoodHashTable::insert_or_update(KeyType key, ValueType value)
 {
-    // TODO
+    ThreadManager manager = this->get_thread_lock_manager();
+    HashCodeType hashcode = this->hash(key);
+    size_t home = this->get_home(hashcode, this->capacity_);
+    size_t idx = home;
+
+    KeyType prev_key = key;
+    ValueType prev_val = value;
+    //bool is_inserted, is_found, found;
+
+    for(size_t distance = 0; this->buckets_[home].offset >= distance; ++distance, ++idx){
+        while(this->buckets_[idx].key == key) {
+            KeyType prev_key_copy = prev_key;
+            ValueType prev_val_copy = prev_val;
+            auto [prev_key, prev_val] = this->compare_and_set_key_val(home, prev_key_copy, this->buckets_[idx].key, prev_val_copy);
```

Comment:
```suggestion
            std::tie(prev_key, prev_val) = this->compare_and_set_key_val(home, key, key, prev_val);
```

1. `std::tie` to simplify.
2. The original algorithm passes the same key and val in every iteration of the loop.